src/parser/expression_parser.cpp

```diff
@@ -141,17 +147,36 @@ std::shared_ptr<expression> parse_expression(const parameter::vector &conditions
             conditions.emplace_back(
                 std::make_unique<exists_negated_condition>(std::move(arguments), limits));
         } else {
-            auto [data_id, matcher] = parse_matcher(operator_name, params);
+            auto raw_operator_name = operator_name;
+            auto negated = operator_name.starts_with('!');
+            if (!negated) {
+                auto [data_id, matcher] = parse_any_matcher(operator_name, params);
 
-            if (!matcher && !data_id.empty()) {
-                data_ids_to_type.emplace(data_id, operator_name);
-            }
+                if (!matcher && !data_id.empty()) {
+                    data_ids_to_type.emplace(data_id, operator_name);
+                }
 
-            auto arguments =
-                parse_arguments<scalar_condition>(params, source, transformers, addresses, limits);
+                auto arguments = parse_arguments<scalar_condition>(
+                    params, source, transformers, addresses, limits);
+                conditions.emplace_back(std::make_unique<scalar_condition>(
+                    std::move(matcher), data_id, std::move(arguments), limits));
+
+            } else {
+                operator_name = operator_name.substr(1);
+                auto [data_id, matcher] =
+                    parse_matcher<matcher::ip_match, matcher::exact_match, matcher::regex_match,
+                        matcher::phrase_match, matcher::equals<>>(operator_name, params);
+
+                if (!matcher && !data_id.empty()) {
+                    data_ids_to_type.emplace(data_id, operator_name);
+                }
 
-            conditions.emplace_back(std::make_unique<scalar_condition>(
-                std::move(matcher), data_id, std::move(arguments), limits));
+                auto arguments = parse_arguments<scalar_negated_condition>(
+                    params, source, transformers, addresses, limits);
+                conditions.emplace_back(
+                    std::make_unique<scalar_negated_condition>(std::move(matcher), data_id,
+                        std::move(arguments), std::string{raw_operator_name}, limits));
+            }
```

Comment:
fwict this could easily be refactored to remove the duplication while improving legibility

Reply:
Done